shared/test/metabase/shared/formatting/date_test.cljc

```diff
@@ -0,0 +1,278 @@
+(ns metabase.shared.formatting.date-test
+  (:require
+   [clojure.test :refer [are deftest is testing]]
+   [metabase.shared.formatting.date :as date]
+   [metabase.shared.util.time :as shared.ut]
+   #?@(:clj [[metabase.shared.formatting.internal.date-formatters :as formatters]
+             [metabase.test :as mt]]))
+  #?(:clj (:import
+           java.util.Locale)))
+
+(deftest format-for-parameter-test
+  (testing "some units have custom formatting"
+    (are [exp date unit] (= exp (date/format-for-parameter (shared.ut/parse-timestamp date) {:unit unit}))
+      ;; Years
+      "2022-01-01~2022-12-31" "2022-12-19T12:03:19" "year"
+      ;; Months
+      "2022-12"               "2022-12-19T12:03:19" "month"
+      "2022-01"               "2022-01-19T12:03:19" "month"
+      ;; Quarters
+      "Q4-2022"               "2022-12-19T12:03:19" "quarter"
+      "Q1-2022"               "2022-01-19T12:03:19" "quarter"
+      ;; Quarter-of-year formats as the same day. This is an odd case, but it's in the original unit tests.
+      ;; It actually isn't recognized as a ".startOf" by Moment, so it doesn't change the input. That means the
+      ;; "range" is the input time, so it renders as a single day reference.
+      "2022-01-19"            "2022-01-19T12:03:19" "quarter-of-year"
+      ;; Days
+      "2022-01-19"            "2022-01-19T12:03:19" "day"
+      "2022-12-19"            "2022-12-19T12:03:19" "day"))
+
+  (testing "other units are treated as days or day ranges"
+    (are [exp date unit] (= exp (date/format-for-parameter (shared.ut/parse-timestamp date)
+                                                           {:unit   unit
+                                                            :locale #?(:cljs nil :clj (Locale/getDefault))}))
+      ;; Hour and minute are treated as days.
+      "2022-12-19" "2022-12-19T12:03:19" "hour"
+      "2022-12-19" "2022-12-19T00:03:19" "hour"
+      "2022-12-19" "2022-12-19T12:03:19" "minute"
+      "2022-12-19" "2022-12-19T00:03:19" "minute"
+
+      ;; Weeks are (previous/this) start of Sunday to end of Saturday
+      "2022-12-18~2022-12-24" "2022-12-18T00:03:19" "week"    ; Sunday
+      "2022-12-18~2022-12-24" "2022-12-19T00:03:19" "week"    ; Monday
+      "2022-12-18~2022-12-24" "2022-12-24T23:03:19" "week"))) ; Saturday
+
+(deftest format-range-with-unit-test
+  (letfn [(week-of [d]
+            (date/format-range-with-unit d {:unit    "week"
+                                            :compact true
+                                            :locale  #?(:cljs nil :clj (Locale/getDefault))}))]
```

Comment:
Is `(Locale/getDefault)` stable enough or should not matter?

Reply:
I believe that's fixed at JVM startup.